oauth2client/contrib/_metadata.py

```diff
@@ -0,0 +1,104 @@
+# Copyright 2016 Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Provides helper methods for talking to the Compute Engine metadata server.
+
+See https://cloud.google.com/compute/docs/metadata
+"""
+
+import datetime
+import httplib2
+import json
+
+from six.moves import http_client
+from six.moves.urllib.parse import urlencode
+
+from oauth2client._helpers import _from_bytes
+from oauth2client.client import _UTCNOW
+
+METADATA_ROOT = 'http://metadata.google.internal/computeMetadata/v1/'
+METADATA_HEADERS = {'Metadata-Flavor': 'Google'}
+
+
+def get(path, http_request=None, root=METADATA_ROOT, **kwargs):
+    if path is None:
+        path = []
+
+    if not http_request:
```

Comment:
So, `gce.AppAssertionCredentials` are going to need at least one http request to fetch information about the service account. In keeping with the JIT http request pattern of the existing class, I don't make that request until `service_account_email` is called. However, if we are comfortable making that request up front it would

(1) Simplify the class a great deal.
(2) Prevent http requests from being made anywhere but the constructor and `_refresh`

Reviewers of #476 went back and forth on this a couple times. So I'd like to make a final decision now =) =)